### PR TITLE
实现 Bitcask ，一种日志结构的哈希表

### LINKED ISSUE #9451: 实现 Bitcask ，一种日志结构的哈希表
* 原文链接：[Implementing Bitcask, a Log-Structured Hash Table](https://healeycodes.com/implementing-bitcask-a-log-structured-hash-table)
* Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/article/2022/implementing-bitcask-a-log-structured-hash-table.md)
* PR 地址：
* 文章分类：**后端**
* **注意：文件位置在 `article/2022` 文件夹中，不是之前的 `article/2021` 了。**
----
* 翻译时间：8 天
* 校对时间：2 天
* 翻译积分：3.5 分
* 校对积分：1.5 分
* [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by wangxuanni**
译文翻译完成，resolve #9451
@lsvih

### CODE REVIEW

**Comment by Quincy-Ye** [comment]
@lsvih 认领校对

**Comment by lsvih** [comment]
@Quincy-Ye 欢迎校对~

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 1]
```diff
@@ -2,29 +2,29 @@
...
+周末，我在一个名为 [bitcask-lite](https://github.com/healeycodes/bitcask-lite) 的项目中实现了 Bitcask 的部分[设计规范](https://riak.com/assets/bitcask-intro.pdf)——一个使用 Go 标准库的键/值数据库和服务器。
 
-I needed to serve some largeish values for a side project and instead of building an MVP with something like SQLite, I [yak shaved](https://seths.blog/2005/03/dont_shave_that/) a database.
+我为一个其他项目提供一些大的值数据，而不是使用 SQLite 之类的东西构建 MVP，我 [yak 了](https://seths.blog/2005/03/dont_shave_that/) 一个数据库。
 
-In bitcask-lite, keys and metadata live in a concurrent map based on [orcaman/concurrent-map](https://github.com/orcaman/concurrent-map) — a map of map shards. Go doesn't allow concurrent reading and writing of maps — so each map shard needs to be locked independently to avoid limiting bitcask-lite's concurrency to a single request.
+在 bitcask-lite 中，键和元数据存在于基于 [orcaman/concurrent-map](https://github.com/orcaman/concurrent-map) 的并发映射中——映射分片的映射。Go 不允许并发的读取和写入映射——因此每个映射分片都需要独立锁定，以避免将 bitcask-lite 的并发限制为单个请求。
```
Comment (Quincy-Ye) [inline article/2022/implementing-bitcask-a-log-structured-hash-table.md:27]:
```suggestion
在 bitcask-lite 中，键和元数据存在于基于 [orcaman/concurrent-map](https://github.com/orcaman/concurrent-map) 的并发映射中——映射分片的映射。Go 不允许并发地读取和写入映射 —— 因此每个映射分片都需要独立锁定，以避免将 bitcask-lite 的并发限制为单个请求。
```

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 2]
```diff
@@ -35,34 +35,34 @@ type MapShard[V any] struct {
 }
 ```
 
-The database is a directory with one or many log files. Items get written to the active log file with the schema: `expire, keySize, valueSize, key, value,` (I'm a big fan of human-readable data).
+这个数据库是包含一个或者多个日志文件的目录。数据使用schema写入活跃日志文件：（`expire, keySize, valueSize, key, value,`我是人类可读数据的忠实拥护者）。
```
Comment (Quincy-Ye) [inline article/2022/implementing-bitcask-a-log-structured-hash-table.md:38]:
```suggestion
这个数据库是包含一个或者多个日志文件的目录。数据使用 schema 写入活跃日志文件：（`expire, keySize, valueSize, key, value,`我是人类可读数据的忠实拥护者）。
```

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 3]
```diff
@@ -35,34 +35,34 @@ type MapShard[V any] struct {
...
 
-An item with a key of `a` and a value of `b` that expires on 10 Aug 2022 looks like this:
+数据键为`a`，值为`b`，于 2022 年 8 月 10 日到期，如下所示：
 
 ```
 1759300313415,1,1,a,b,
 ```
 
-Log files are append-only and don't need to be locked for **getting** values but when **setting** values there's a lock on the active log file to ensure that the database is correct relative to the order of incoming requests. Unfortunately, this means that write-heavy workloads will perform worse than read-heavy ones.
+日志文件是 append-only （只许追加），所以不需要为*获取*值而上锁， 但是当*设置*值时，会为活跃日志文件上一个锁，以确保数据库相对于传入请求的顺序是正确的。不幸地是，这意味着重写的负载将比重读的负载表现更差。
```
Comment (Quincy-Ye) [inline article/2022/implementing-bitcask-a-log-structured-hash-table.md:46]:
```suggestion
日志文件是 append-only （只许追加），所以不需要为获取值而上锁， 但是当设置值时，会为活跃日志文件上一个锁，以确保数据库相对于传入请求的顺序是正确的。不幸地是，这意味着重写的负载将比重读的负载表现更差。
```

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 4]
```diff
@@ -99,7 +99,7 @@ The trickiest part of this project was parsing the log files; largely due to off
 
 ## HTTP API
 
-I started this project with the API sketched out on a single sticky note.
+我在一张便利贴上勾勒出 API 来开始这个项目。
```
Comment (Quincy-Ye) [inline article/2022/implementing-bitcask-a-log-structured-hash-table.md:102]:
```suggestion
我在一张便利贴上草拟了一下 API 就开始了这个项目。
```

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 5]
```diff
@@ -109,38 +109,38 @@ I started this project with the API sketched out on a single sticky note.
   - expire is optional (default is infinite)
 ```
 
-After finishing it, I saw there were more lines of test code than other code — which says a lot about software complexity in general. I can describe this API in a single breath but it took me a good few hours to cover every edge case with tests and fixtures.
+完成之后，我看到测试代码的行数超过了其他的代码。这大体上说明了软件的复杂度。我可以一口气描述这个 API，但我花了好几个小时来用测试和修复每个边缘情况。
```
Comment (Quincy-Ye) [inline article/2022/implementing-bitcask-a-log-structured-hash-table.md:112]:
```suggestion
完成之后，我看到测试代码的行数超过了其他的代码。这大体上说明了软件的复杂度。我可以一口气描述这个 API，但我花了好几个小时来用测试和修复来覆盖每个边界值用例。
```

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 6]
```diff
@@ -149,9 +149,9 @@ if err != nil {
 }
 ```
 
-I'm happy with the shortcuts I took. And so far, my toy database has been humming along just fine.
+我对我走的捷径很满意。到目前为止，我的玩具数据库一直运转良好。
```
Comment (Quincy-Ye) [inline article/2022/implementing-bitcask-a-log-structured-hash-table.md:152]: 这里的 toy，应该是指的作者写着玩的数据库。直译成玩具数据库，有点奇怪。
```suggestion
我对我走的捷径很满意。到目前为止，我的用来测试的数据库一直运转良好。
```

**Comment by Quincy-Ye** [comment]
@wangxuanni @lsvih 校验完成，若有疑问，随时沟通~

**Comment by CompetitiveLin** [comment]
@lsvih 认领校对

**Comment by lsvih** [comment]
@CompetitiveLin 欢迎校对~

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md (hunk 6 above)
Comment (wangxuanni) [reply article/2022/implementing-bitcask-a-log-structured-hash-table.md:152]: 确实如此，但是翻译这个数据库是用来测试的这层意思也不太符合原文，那么翻译成：我的”玩具“数据库一直运转良好。 如何？

**Comment by wangxuanni** [comment]
感谢校对，已提交修改~

**Review by CompetitiveLin — COMMENTED** [review]
校对完成，最终以译者为准。

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md (hunk 6 above)
Comment (CompetitiveLin) [reply article/2022/implementing-bitcask-a-log-structured-hash-table.md]: 个人认为，”玩具“数据库没问题。另外，作者的这个数据库不是用来测试的吧，是自己实现的一个数据库。

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md [hunk 7]
```diff
@@ -149,9 +149,9 @@ if err != nil {
 }
 ```
 
-I'm happy with the shortcuts I took. And so far, my toy database has been humming along just fine.
+我对我走的捷径很满意。到目前为止，我的“玩具”数据库一直运转良好。
 
-Sure, using SQLite with a single table would have gotten me up and going much quicker. But I mostly took this database detour for fun. I do wonder about the performance comparison between [bitcask-lite](https://github.com/healeycodes/bitcask-lite) and SQLite+server. I was about to set up some benchmarks when I couldn't figure out if you can stream individual SQLite values to a client. If you know, [let me know](mailto:healeycodes@gmail.com)!
+当然，将 SQLite 与单个表一起使用会更快。但我主要是为了好玩写这个数据库。[我确实想知道 bitcask-lite ](https://github.com/healeycodes/bitcask-lite)和 SQLite+server 之间的性能比较。我要设置一些基准时，无法弄清楚是否可以将单个 SQLite 值流式传输到客户端。如果你知道，请[告诉我](mailto:healeycodes@gmail.com)！
```
[the remark this answers is not in the document]
Comment (wangxuanni) [reply article/2022/implementing-bitcask-a-log-structured-hash-table.md:154]: 其他欣然接受，但认为从词意和语境上，comparison应译为比较。

File: article/2022/implementing-bitcask-a-log-structured-hash-table.md (hunk 6 above)
Comment (Quincy-Ye) [reply article/2022/implementing-bitcask-a-log-structured-hash-table.md]: 作者实现这个数据库，就是为了测试性能间的差距。这就是我翻译成测试数据库的原因。原文中作者也有解释到。他是真的好奇性能之间的差距。

**Review by lsvih — DISMISSED** [review]

**Comment by wangxuanni** [comment]
已发布，已投稿~ 
掘金https://juejin.cn/post/7174345557861728292 
知乎https://zhuanlan.zhihu.com/p/590022272

**Comment by lsvih** [comment]
收到！
